torch/_inductor/fx_passes/mkldnn_fusion.py

```diff
@@ -655,17 +655,36 @@ def _recover_linear():
         def reshape_linear_reshape_pattern(match, *args, **kwargs):
             reshape_1 = kwargs.get("reshape_1")
             reshape_2 = kwargs.get("reshape_2")
+            assert len(reshape_1) == 2
+            dynamic_shapes = not all(
+                isinstance(x, int) for x in ([reshape_1[0]] + reshape_2[:-1])
+            )
+
             graph = match.graph
-            node = match.output_node()
-            if reshape_1[0] == reduce(lambda x, y: x * y, reshape_2[:-1]):
+            reshape_2_node = match.output_node()
+            linear_input_node = reshape_2_node.args[0].args[0].args[0]
+            # check linear's input's shape[:-1] == reshape_2[:-1]
+            # and check product(reshape_2[:-1]) == reshape_1[0]
+            if dynamic_shapes:
+                # TODO: Haozhe investigate how add guard here
+                return
+            else:
+                can_remove_reshape = linear_input_node.meta.get("val").shape[
+                    :-1
+                ] == torch.Size(reshape_2[:-1])
+                can_remove_reshape = can_remove_reshape and (
+                    reduce(lambda x, y: x * y, reshape_2[:-1]) == reshape_1[0]
+                )
```

Comment:
How about this. It doesn't introduce a guard but can still handle simple symbolically equivalent cases:
```python
            can_remove_reshape = (
                V.graph.sizevars.statically_known_list_equals(
                    linear_input_node.meta.get("val").shape[:-1],
                    reshape_2[:-1]
                )
                and V.graph.sizevars.statically_known_equals(
                    sympy_product(reshape_2[:-1], reshape_1[0]
                )
            )
```

(Assuming V.graph is available at this point?)

Reply:
Hi, @peterbell10. `V.graph` is not available at this point.
I tried some code from `shape_env.evaulte_expr`, but does not quite understand it yet.
And since it will bring obvious performance regression for https://github.com/pytorch/pytorch/pull/106300.
Can we first disable this for dynamic case?

Reply:
Yeah disabling is fine, just thought this would be nicer. If you have a `shape_env` then you could use `shape_env._maybe_evaluate_static` directly?

Reply:
Thanks for the advice, I will continue to investigate this